deploy/kubernetes/jupyterhub-config.py

```diff
@@ -0,0 +1,164 @@
+import os,sys,subprocess
+import requests
+from oauthenticator.generic import LocalGenericOAuthenticator
+from kubespawner import KubeSpawner
+
+class DemoFormSpawner(KubeSpawner):
+    def __init__(self):
+        self.stacks_path = '/stacks'
+
+        # Filename for the base stack pasted from Jenkins pipeline
+        self.base = 'BASE_STACK_VALUE'
+
+        # Groups of stacks filenames and their names are pasted from Jenkins pipeline
+        # Format:
+        #   self.stacks = [['stack1.yaml', 'stack2.yaml'], ['stack3.yaml'], ['stack4.yaml']]
+        #   self.stacks_names = ['Name A', 'Name B', 'Name C']
+        # Lengths of self.stacks and self.stacks_names must be the same
+        self.stacks = STACKS_VALUE
+        self.stacks_names = STACKS_NAMES_VALUE
+        
+        # Prepend path to stacks folder
+        self.base = os.path.join(self.stacks_path, self.base)
+        self.stacks = [[os.path.join(self.stacks_path, s) for s in stack] for stack in self.stacks]
+
+        self.options = [f'option{i+1}' for i in range(len(stacks))]
+        
+        self.form = """
+<div class="form-group">
+    Choose language kernels and/or packages:
+    <br>"""
+        
+        for i in range(len(stacks)):
+            self.form +=(f"""
+    <input type="checkbox" id="option{i+1}" name="option{i+1}">
+    <label for="option1">{self.stacks_names[i]}</label><br>""")
+        
+        self.form += """
+</div>"""
+
+    def _options_form_default(self):
+        # Generate html form
+        return self.form
+
+    def options_from_form(self, formdata):
+        # Decode user choices        
+        options = [True if formdata.get(option, None) else False for option in self.options]
+        
+        # Get image hash by running `polus-railyard` 
+        tag = subprocess.run(('railyard hash ' + '-b ' + base + ' ' + ' '.join([f'-a {item}' for stack,included in zip(self.stacks,options) for item in stack if included])).split(' '), capture_output=True).stdout.decode("utf-8").rstrip()
+        
+        # Get full image tag
+        image = 'labshare/polyglot-notebook:' + tag
+
+        #Check if image exist to avoid PullErrors
+        if requests.get('https://hub.docker.com/v2/repositories/labshare/polyglot-notebook/tags/' + tag).status_code != 200:
+            #if image does not exist, switch to the maximal image with all stacks included
+            self.log.debug("Requested image %s is not in registry, using default image", image)
+            options = [True] * len(stacks)
+            tag = subprocess.run(('railyard hash ' + '-b ' + base + ' ' + ' '.join([f'-a {item}' for stack,included in zip(stacks,options) for item in stack if included])).split(' '), capture_output=True).stdout.decode("utf-8").rstrip()
+            image = 'labshare/polyglot-notebook:' + tag
+            
+        setattr(self, 'image', image)
+        return dict(profile={'display_name': 'Chosen image', 'default': True, 'kubespawner_override': {'image': image,}})
```

Comment:
I actually meant if we could separate this specific portion out as a separate file, and probably add it to the JupyterHub Docker image. It serves as a piece of functionality, distinct from the configurations below. Can we also give the class a different name? Your code isn't demo code, it's the first iteration of a long process, something informative like `ModularSpawner`.